File: packages/python/cronixui/chip.py
"""CronixUI Chip Component - Native tkinter implementation."""

from __future__ import annotations
import tkinter as tk
from tkinter import ttk
from typing import Callable, Optional
from .core import Frame, get_theme


class Chip(ttk.Label):
    """Native chip/tag element with optional remove button.

    Args:
        parent: Parent widget
        text: Chip text
        variant: Chip variant (default, accent, success, warning, error, info)
        size: Chip size (sm, md, lg)
        selected: Whether chip is selected
        removable: Show remove button
        on_click: Click handler
        on_remove: Remove handler

    Example:
        >>> chip = Chip(root, "Option A", variant="accent", selected=True)
        >>> chip.pack()
    """

    VARIANTS = ("default", "accent", "success", "warning", "error", "info")

    def __init__(
        self,
        master: tk.Misc,
        text: str = "",
        variant: str = "default",
        size: str = "md",
        selected: bool = False,
        removable: bool = False,
        on_click: Optional[Callable] = None,
        on_remove: Optional[Callable] = None,
        **kwargs
    ):
        self.variant = variant
        self.selected = selected
        theme = get_theme()

        # Determine colors
        if selected:
            bg, fg = theme.accent, theme.text
        elif variant == "success":
            bg, fg = "#1a3a1a", "#4a8b4a"
        elif variant == "warning":
            bg, fg = "#3a2a0a", "#8b7a4a"
        elif variant == "error":
            bg, fg = "#3a1a1a", theme.error
        elif variant == "info":
            bg, fg = "#1a2a3a", "#4a6a8b"
        elif variant == "accent":
            bg, fg = theme.accent_light, theme.accent
        else:
            bg, fg = theme.surface_2, theme.text

        font_size = {"sm": 9, "md": 10, "lg": 12}.get(size, 10)

        display_text = f" {text} ✕" if removable else text
        super().__init__(master, text=display_text, background=bg, foreground=fg,
                         font=(theme.font_family, font_size), padding=(8, 2), **kwargs)

        if on_click:
            self.bind('<Button-1>', lambda e: on_click())
            self.configure(cursor='hand2')
        if on_remove and removable:
            self.bind('<Button-1>', lambda e: on_remove())

    def set_selected(self, selected: bool) -> None:
        """Update selected state."""
        self.selected = selected
        theme = get_theme()
        if selected:
            self.configure(background=theme.accent, foreground=theme.text)
        else:
            self.configure(background=theme.surface_2, foreground=theme.text)

File: packages/rust/cronixui/src/components/chip.rs
//! Chip component

/// Chip variant.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ChipVariant {
    Default,
    Accent,
    Success,
    Warning,
    Error,
    Info,
}

/// Chip/tag with optional remove button.
///
/// ## Example
/// ```no_run
/// use cronixui::components::{Chip, ChipVariant};
///
/// let chip = Chip::new("Option A")
///     .with_variant(ChipVariant::Accent)
///     .selected(true);
/// assert!(chip.is_selected());
/// ```
pub struct Chip {
    pub label: String,
    pub variant: ChipVariant,
    pub selected: bool,
    pub removable: bool,
}

impl Chip {
    pub fn new(label: &str) -> Self {
        Self { label: label.to_string(), variant: ChipVariant::Default, selected: false, removable: false }
    }

    pub fn with_variant(mut self, variant: ChipVariant) -> Self {
        self.variant = variant;
        self
    }

    pub fn selected(mut self, selected: bool) -> Self {
        self.selected = selected;
        self
    }

    pub fn removable(mut self, removable: bool) -> Self {
        self.removable = removable;
        self
    }

    pub fn is_selected(&self) -> bool {
        self.selected
    }

    /// Render as text.
    pub fn render_text(&self) -> String {
        if self.removable {
            format!("{} ✕", self.label)
        } else {
            self.label.clone()
        }
    }
}
